File: src/main/java/app/RunConfiguration.java
package app;

import org.apache.commons.cli.*;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;

import org.json.*;



public class RunConfiguration {

    // static file locations in JAR
    public static final String DEFAULT_CONFIG_FILE_PATH = "/.conf.json";
    public static final String USER_CONFIG_FILE_PATH = "/user.conf.json";

    // static final locations not in JAR
    public static final String APP_DATA_DIR = "filevault";
    public static final String CACHE_DIR = "tmp_" + String.valueOf(System.currentTimeMillis());
    public static final String APP_ARCHIVE_FILE = "archive.dat";

    // Settings
    private HashMap<String, String> configMap;
    private String[] commandLineArgs;

    public long sessionID;

    public RunConfiguration() {
        this.sessionID = System.currentTimeMillis();
        this.setupConfig();
    }

    private void setupConfig() {
        this.configMap = new HashMap<>();
        this.configMap.put("appName", App.APP_NAME);
        this.configMap.put("appNamePretty", App.APP_NAME_PRETTY);
        this.configMap.put("javaVersion", System.getProperty("java.version"));
        this.configMap.put("javafxVersion", System.getProperty("javafx.version"));
        this.configMap.put("userHome", System.getProperty("user.home"));
        this.configMap.put("runPath", new File(".").getAbsolutePath());
    }

    public String getAppName() { return this.configMap.get("appName"); }

    public String getAppNamePretty() { return this.configMap.get("appNamePretty"); }

    public long getSessionID() { return this.sessionID; }

    public String updateConfig(String key, String value, boolean persist) {
        if(!RunConfigurationConstraints.checkConstraints(key, value)) {
            throw new IllegalArgumentException("Invalid config values");
        }
        String old = this.configMap.put(key, value);

        if(persist) {
            // TODO write to user settings file
        }

        return old;
    }

    public String getAppHomePathAbsolute() {
        return this.configMap.get("userHome") +
                File.separator + RunConfiguration.APP_DATA_DIR;
    }

    public String getCachePathAbsolute() {
         return this.getAppHomePathAbsolute() + File.separator + RunConfiguration.APP_ARCHIVE_FILE;
    }

    public String getArchivePathAbsolute() {
        return this.getAppHomePathAbsolute() + File.separator + RunConfiguration.APP_ARCHIVE_FILE;
    }

    // TODO Make sure to validate all user generated config
    public void parseCommandLineArgs(String[] arguments) throws ParseException {
        this.commandLineArgs = arguments;

        Options opts = new Options();

        Option gui = new Option("g", "gui", false, "uses a GUI");
        gui.setRequired(false);
        opts.addOption(gui);

        Option help = new Option("h", "help", false, "print the usage text");
        gui.setRequired(false);
        opts.addOption(help);

        CommandLineParser parser = new DefaultParser();
        HelpFormatter formatter = new HelpFormatter();
        CommandLine cmd = null;

        try {
            cmd = parser.parse(opts, arguments);
        } catch (ParseException e) {
            formatter.printHelp(this.getConfigString("appName"), opts);
            throw e;
        }

        if(cmd.hasOption(help)) {
            formatter.printHelp(this.getConfigString("appName"), opts);
            System.exit(0);
        }

        this.configMap.put("gui", String.valueOf(cmd.hasOption(gui)));
    }

    // TODO Make sure to validate all user generated config
    public void parseUserConfigFile() throws IOException {

        InputStream in = RunConfiguration.class.getResourceAsStream(RunConfiguration.USER_CONFIG_FILE_PATH);
        JSONObject json = new JSONObject(new String(in.readAllBytes()));

        for(String k : JSONObject.getNames(json)) {

            // do not add extra config keys
            if(!this.configMap.containsKey(k)) { continue; }

            String v = String.valueOf(json.get(k));
            if(k == null || v == null) { continue; }

            // config value constraint checking
            if(!RunConfigurationConstraints.checkConstraints(k, v)) { continue; }

            this.configMap.put(k, v);
        }

    }

    public void parseDefaultConfigFile() throws IOException {

        InputStream in = RunConfiguration.class.getResourceAsStream(RunConfiguration.DEFAULT_CONFIG_FILE_PATH);
        JSONObject json = new JSONObject(new String(in.readAllBytes()));

        for(String k : JSONObject.getNames(json)) {
            this.configMap.put(k, String.valueOf(json.get(k)));
        }

    }

    public boolean contains(String key) {
        return this.configMap.containsKey(key);
    }

    public void reloadConfig() throws ParseException, IOException {
        this.parseDefaultConfigFile();
        this.parseUserConfigFile();
        this.parseCommandLineArgs(this.commandLineArgs);
    }

    public void resetConfig() throws IOException {
        this.setupConfig();
    }

    public int getConfigInt(String key) {
        return Integer.parseInt(this.configMap.get(key));
    }

    public boolean getConfigBool(String key) {
        return Boolean.parseBoolean(this.configMap.get(key));
    }

    public String getConfigString(String key) {
        return this.configMap.get(key);
    }

    public double getConfigDouble(String key) {
        return Double.parseDouble(this.configMap.get(key));
    }

    public char getConfigChar(String key) {
        String v = this.configMap.get(key);
        if(v == null || v.length() == 0) {
            return (char) 0;
        } else {
            return this.configMap.get(key).charAt(0);
        }
    }

    public long getConfigLong(String key) {
        return Long.parseLong(this.configMap.get(key));
    }

    public String toString() {
        // TODO implement pretty string
        return "Run Configuration:";
    }

    private static class RunConfigurationConstraints {

        // configuration constraints
        private static final String INT_FIELDS = ",fontSizeLG,fontSizeMD,fontSizeSM,";
        private static final String POS_INT_FIELDS = ",fontSizeLG,fontSizeMD,fontSizeSM,";
        private static final String DOUBLE_FIELDS = "";
        private static final String STRING_FIELDS = ",cliPrefix,";
        private static final String CHAR_FIELDS = "";
        private static final String BOOL_FIELDS = ",gui,darkMode,";

        private static boolean checkConstraints(String key, String value) {
            String pKey = "," + key + ",";

            if(RunConfigurationConstraints.INT_FIELDS.contains(pKey)) {
                try { Integer.parseInt(value); }
                catch(NumberFormatException | NullPointerException e) { return false; }
            }

            if(RunConfigurationConstraints.POS_INT_FIELDS.contains(pKey)) {
                try {
                    int i = Integer.parseInt(value);
                    if(i <= 0) { return false; }
                }
                catch(NumberFormatException | NullPointerException e) { return false; }
            }

            if(RunConfigurationConstraints.DOUBLE_FIELDS.contains(pKey)) {
                try { Double.parseDouble(value); }
                catch(NumberFormatException | NullPointerException e) { return false; }
            }

            if(RunConfigurationConstraints.STRING_FIELDS.contains(pKey)) {
                return value != null;
            }

            if(RunConfigurationConstraints.CHAR_FIELDS.contains(pKey)) {
                return value != null && value.length() == 1;
            }

            if(RunConfigurationConstraints.BOOL_FIELDS.contains(pKey)) {
                return value.equalsIgnoreCase("true") || value.equalsIgnoreCase("false");
            }

            return true;
        }


    }

}

File: rs/src/app/run_configuration.rs
use clap::{Arg, ArgAction, Command};
use std::collections::HashMap;
use std::env::Args;
use std::fmt;
use std::fs;

use crate::exceptions::config_parse_error::ConfigParseError;

#[derive(Clone)]
pub struct RunConfiguration {
    config_map: HashMap<String, String>,
    session_id: u64,
    args: Vec<String>,
}

// static project information
const GITHUB_WEB_PATH: &'static str = "https://github.com/sameer-n012/tag-vfs";
const PROJECT_WEB_PATH: &'static str = "";
const APP_NAME: &'static str = "file-vault";
const APP_NAME_PRETTY: &'static str = "File Vault";

// static file locations in JAR
const DEFAULT_CONFIG_FILE_PATH: &'static str = "/.conf.json";
const USER_CONFIG_FILE_PATH: &'static str = "/user.conf.json";

// static final locations not in JAR
const APP_DATA_DIR: &'static str = "filevault";
const CACHE_DIR: &'static str = "tmp_";
const APP_ARCHIVE_FILE: &'static str = "archive.dat";

impl RunConfiguration {
    pub fn new(args: Args) -> Self {
        let mut config = RunConfiguration {
            config_map: HashMap::new(),
            session_id: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_secs(),
            args: Vec::from_iter(args),
        };
        config.setup_config();
        config
    }

    pub fn get_app_name(&self) -> String {
        return self
            .config_map
            .get("appName")
            .unwrap_or(&APP_NAME.to_string())
            .to_string();
    }

    pub fn get_app_name_pretty(&self) -> String {
        return self
            .config_map
            .get("appNamePretty")
            .unwrap_or(&APP_NAME_PRETTY.to_string())
            .to_string();
    }

    pub fn get_session_id(&self) -> u64 {
        return self.session_id;
    }

    fn setup_config(&mut self) {
        self.config_map
            .insert("appName".to_string(), APP_NAME.to_string());
        self.config_map
            .insert("appNamePretty".to_string(), APP_NAME_PRETTY.to_string());
        self.config_map.insert(
            "runPath".to_string(),
            std::env::current_dir()
                .unwrap_or_default()
                .display()
                .to_string(),
        );
    }

    pub fn update_config(&mut self, key: String, value: String, persist: bool) -> String {
        if !RunConfiguration::check_config_constraints(&key) {
            print!("{}", ConfigParseError::new("Invalid config value"));
        }

        let old = self.config_map.insert(key, value).unwrap_or("".to_string());
        if persist {
            // TODO write to file
        }
        return old;
    }

    pub fn parse_command_line_args(&mut self) -> Result<(), String> {
        let matches = Command::new("tvfs")
            .version("1.0")
            .about("A tag-based file system")
            .arg(
                Arg::new("gui")
                    .short('g')
                    .long("gui")
                    .help("uses the GUI")
                    .required(false)
                    .action(ArgAction::SetTrue),
            )
            .arg(
                Arg::new("help")
                    .short('h')
                    .long("help")
                    .help("prints the usage text")
                    .required(false)
                    .action(ArgAction::Help),
            )
            .try_get_matches_from(self.args.clone())
            .unwrap_or_default();

        if matches.contains_id("gui") {
            self.config_map
                .insert("gui".to_string(), "true".to_string());
        }

        if matches.contains_id("help") {
            std::process::exit(0);
        }

        Ok(())
    }

    pub fn parse_default_config_file(&mut self) -> Result<(), String> {
        let config_data = fs::read_to_string("resources/.conf.json").map_err(|e| e.to_string())?;
        let config_json: serde_json::Value =
            serde_json::from_str(&config_data).map_err(|e| e.to_string())?;
        if let serde_json::Value::Object(map) = config_json {
            for (key, value) in map {
                self.config_map.insert(key, value.to_string());
            }
        }
        return Ok(());
    }

    pub fn parse_user_config_file(&mut self) -> Result<(), String> {
        let config_data =
            fs::read_to_string("resources/user.conf.json").map_err(|e| e.to_string())?;
        let config_json: serde_json::Value =
            serde_json::from_str(&config_data).map_err(|e| e.to_string())?;
        if let serde_json::Value::Object(map) = config_json {
            for (key, value) in map {
                if !RunConfiguration::check_config_constraints(&key) {
                    continue;
                }

                if !self.config_map.contains_key(&key) {
                    continue;
                }

                self.config_map.insert(key, value.to_string());
            }
        }

        return Ok(());
    }

    pub fn get_app_home_path_absolute(&self) -> String {
        format!("{}/filevault", dirs::home_dir().unwrap().display())
    }

    pub fn get_cache_path_absolute(&self) -> String {
        format!(
            "{}/tmp_{}",
            self.get_app_home_path_absolute(),
            self.session_id
        )
    }

    pub fn get_archive_path_absolute(&self) -> String {
        format!("{}/archive.dat", self.get_app_home_path_absolute())
    }

    pub fn contains(&self, key: &str) -> bool {
        return self.config_map.contains_key(key);
    }

    pub fn reload_config(&mut self) {
        self.parse_default_config_file();
        self.parse_user_config_file();
        self.parse_command_line_args();
    }

    pub fn reset_config(&mut self) {
        self.setup_config();
    }

    pub fn get_config_int(&self, key: &str) -> i32 {
        return self
            .config_map
            .get(key)
            .and_then(|value| value.parse::<i32>().ok())
            .unwrap_or_default();
    }

    pub fn get_config_long(&self, key: &str) -> i64 {
        return self
            .config_map
            .get(key)
            .and_then(|value| value.parse::<i64>().ok())
            .unwrap_or_default();
    }

    pub fn get_config_double(&self, key: &str) -> f64 {
        return self
            .config_map
            .get(key)
            .and_then(|value| value.parse::<f64>().ok())
            .unwrap_or_default();
    }

    pub fn get_config_bool(&self, key: &str) -> bool {
        return self
            .config_map
            .get(key)
            .and_then(|value| value.parse::<bool>().ok())
            .unwrap_or_default();
    }

    pub fn get_config_char(&self, key: &str) -> String {
        return self
            .config_map
            .get(key)
            .map(|value| value.chars().next().unwrap_or('\0').to_string())
            .unwrap_or_else(|| '\0'.to_string());
    }

    pub fn get_config_string(&self, key: &str) -> String {
        self.config_map
            .get(key)
            .unwrap_or(&"".to_string())
            .to_string()
    }

    // Run configuration constraints
    const INT_FIELDS: [&'static str; 3] = ["fontSizeLG", "fontSizeMD", "fontSizeSM"];
    const POS_INT_FIELDS: [&'static str; 3] = ["fontSizeLG", "fontSizeMD", "fontSizeSM"];
    const DOUBLE_FIELDS: [&'static str; 0] = [];
    const STRING_FIELDS: [&'static str; 1] = ["cliPrefix"];
    const CHAR_FIELDS: [&'static str; 0] = [];
    const BOOL_FIELDS: [&'static str; 2] = ["gui", "darkMode"];

    fn check_config_constraints(key: &str) -> bool {
        if RunConfiguration::INT_FIELDS.contains(&key) {
            // try to parse as int, if not return falsereturn key
            if !key.parse::<i32>().is_ok() {
                return false;
            }
        } else if RunConfiguration::POS_INT_FIELDS.contains(&key) {
            if !key.parse::<i32>().is_ok_and(|x| x > 0) {
                return false;
            }
        } else if RunConfiguration::DOUBLE_FIELDS.contains(&key) {
            if !key.parse::<f64>().is_ok() {
                return false;
            }
        } else if RunConfiguration::STRING_FIELDS.contains(&key) {
            // Any string is valid
        } else if RunConfiguration::CHAR_FIELDS.contains(&key) {
            if key.chars().count() != 1 {
                return false;
            }
        } else if RunConfiguration::BOOL_FIELDS.contains(&key) {
            if !key.parse::<bool>().is_ok() {
                return false;
            }
        }
        true
    }
}

impl fmt::Display for RunConfiguration {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        // TODO pretty print
        write!(f, "Run Configuration")
    }
}
